src/librustdoc/clean/mod.rs

```diff
@@ -1152,12 +1152,31 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
             }
         }
         ty::AssocKind::Fn => {
-            let generics = clean_ty_generics(
+            let sig = tcx.fn_sig(assoc_item.def_id);
+
+            // FIXME: Remove this once #103448 has been merged.
+            let mut late_bound_regions: Vec<_> = cx
+                .tcx
+                .collect_referenced_late_bound_regions(&sig)
+                .into_iter()
+                .filter_map(|br| match br {
+                    ty::BrNamed(_, name) if name != kw::UnderscoreLifetime => Some(name),
+                    _ => None,
+                })
+                .collect();
+
+            // Sort late-bound lifetimes alphabetically to guarantee a stable and deterministic output.
```

Comment:
Why is this required? The output of `fn_sig` should already be deterministic.

Reply:
Maybe *stable* & *deterministic* is not the correct choice of words here. I mostly mean *predictable*.
I am sure `fn_sig` is deterministic but I am more worried about `collect_referenced_late_bound_regions` which returns a `FxHashSet` and which I think is the culprit:

Assuming I don't sort the late-bound lifetimes alphabetically, then their order in following code snippet depends on the shape of the rest of the code. If I comment out the methods `T::a`–`T::l`, then the parameter list of `T::test` is rendered as `<'a, 'b, 'c>` (on my machine at least). If I get rid of the comment again, then it is rendered as `<'c, 'a, 'b>` (in this case). Changing the name of `T::test` (e.g. to `sep`) also changes the order around.

<details>
<summary>code</summary>

```rust
pub struct T;
impl T {
    pub fn a<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn b<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn c<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn d<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn e<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn f<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn g<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn h<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn i<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn j<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn k<'x>(_: fn(&'x ()) -> &'x ()) {}
    pub fn l<'x>(_: fn(&'x ()) -> &'x ()) {}

    pub fn test<'a, 'b, 'c>(_: &'a &'b str, _: &(dyn ToString + 'c)) {}
}
```

</details>

Reply:
I am going to update all those comments to use the word *predictable* instead and to mention the hash set.